Enable aten-aten decomps

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #86480
* __->__ #85921

Invokes aten-aten decomps with re-entrant FakeMode. These decomps are being used in other places, so it's good to unify the path static fake tensor takes / get additional testing etc. There is also an instance where we return different devices with cpu/cuda which this fixes ([batch_norm](https://github.com/pytorch/pytorch/blob/master/torch/_decomp/decompositions.py#L1374))

eellison committed:
Enable aten-aten decomps

[ghstack-poisoned]

eellison committed:
Update on "Enable aten-aten decomps"




[ghstack-poisoned]

eellison committed:
Update on "Enable aten-aten decomps"




[ghstack-poisoned]

eellison committed:
Update on "Enable aten-aten decomps"




[ghstack-poisoned]

eellison committed:
Update on "Enable aten-aten decomps"




[ghstack-poisoned]

eellison committed:
Update on "Enable aten-aten decomps"


Invokes aten-aten decomps with re-entrant FakeMode. These decomps are being used in other places, so it's good to unify the path static fake tensor takes / get additional testing etc. There is also an instance where we return different devices with cpu/cuda which this fixes ([batch_norm](https://github.com/pytorch/pytorch/blob/master/torch/_decomp/decompositions.py#L1374))



[ghstack-poisoned]

eellison committed:
Update on "Enable aten-aten decomps"


Invokes aten-aten decomps with re-entrant FakeMode. These decomps are being used in other places, so it's good to unify the path static fake tensor takes / get additional testing etc. There is also an instance where we return different devices with cpu/cuda which this fixes ([batch_norm](https://github.com/pytorch/pytorch/blob/master/torch/_decomp/decompositions.py#L1374))



[ghstack-poisoned]

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -115,6 +115,19 @@ def get_schema_info(func):
+    from torch._decomp import decomposition_table
+
+    decompositions = torch._decomp.decompositions
+    decomp_attrs = [getattr(decompositions, attr) for attr in dir(decompositions)]
```
dir? 🤔

eellison replied:
do we not like dir ? alternatively i could define a register_decomposition in `torch?/_decomp/decompositions` that adds these ops to a list

ezyang replied:
It's just weird to do it this way because, like, we explicitly have 

```
# None of these functions are publicly accessible; get at them
# from torch._decomps
__all__: List[str] = []
```

so like why aren't you getting the decomps from the decompositions table lol

ezyang replied:
The answer being, that, you ONLY want the decomps that were registered inside torch._decomp.decompositions, as those "tend" to be aten-to-aten decompositions. But it seems more semantically correct for this to be some sort of attribute associated with any given decomp (you could even compute it automatically, by checking if the free variables of the code object reference any prims); after all, you're going to miss aten-to-aten decomps that just happen to be in torch._refs.

I can still be convinced that this is a good unblock, but the commentary on the function isn't enough to figure out the nuance.

eellison replied:
```
# many of the prims do not at the moment model aliasing or strides,
# but as an incremental step, enable aten-aten decomp. these decomps
# are used for aot autograd tracing so we would like to unify and add
# additional testing to them
``` 

was the comment - I think the misleading thing here is just the `aten-aten` name

eellison replied:
> so like why aren't you getting the decomps from the decompositions table lol
This seemed like the smallest change set and least invasive for what is supposed to be a somewhat temporary state. All of the decomps will eventually need to work 100% accurately for dynamic shapes, although we're a bit away from that at the moment. I'm happy to take a different route to accomplish that if you have another suggestion.

As for why I am enabling just these decomps - these are the decomps that are being enabled for torchinductor, not all of the prims, so they are not on the critical path for correctness with static shapes. For the ones that are (the ones I am using here), I think it's important to consolidate on their implementation, and add additional testing. For instance, it would have caught https://github.com/pytorch/torchdynamo/issues/1356.

eellison replied:
I'm not sure what's a better name than `aten_to_aten`. `non_prim_decomp` ?

ezyang replied:
I guess, this isn't really something you resolve with naming; it's something you resolve with comments, and then socializing the comments with people. For example, we still have @mruberry telling folks that "eventually, we should move all torch._decomp to torch._refs". But this change is antithetical to that idea, since moving a decomp from here to refs prevents it from being exercised here. I don't think the idea here is fundamentally incompatible, but there is some needle to thread here.

ezyang replied:
You could also name this exactly what it does on the tin: `torch_decomp_decompositions`

bdhirsh commented on torch/csrc/Module.cpp:
```diff
@@ -1342,17 +1342,15 @@ Call this whenever a new thread is created in order to propagate values from
+      local_keyset.included_ = local_keyset.included_ | key_set;
+    } else {
+      auto k = key_set.highestBackendKey();
+      local_keyset.included_ = local_keyset.included_.remove_backend(k);
```
(same nit as the previous PR, but you can just do `.remove_backend(c10::BackendComponent::MetaBit)` here)

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -769,6 +783,15 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
                     return r
 
+        # invoke aten-aten decomps
+        with self:
```
better to avoid reentering the context manager until we know we're calling into the decomp?

ezyang commented on torch/_decomp/decompositions.py:
```diff
@@ -1634,7 +1630,7 @@ def maybe_mask(vals, length, range_max, adaptive, dim):
 
 
-@register_decomposition(aten.index_add_)
+@register_decomposition(aten.index_add_, disable_meta=True)
```
It would be good to have some commentary about why we switched this. IIRC, @Chillee said it's because `index_add_` decomp is wrong?

ezyang commented on torch/_decomp/__init__.py:
```diff
@@ -18,6 +18,8 @@
 
 meta_lib = torch.library.Library("aten", "IMPL", "Meta")
 
+_disabled_meta_decomps = set()
```
comment?

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -366,7 +380,7 @@ def run_and_return_new_tensor_of_input_device(fake_mode, func, args, kwargs):
 
     out_device = new_kwargs["input"].device
-    with in_kernel_invocation_manager(fake_mode):
+    with no_dispatch(), in_kernel_invocation_manager(fake_mode):
```
So is the invariant now that you always `no_dispatch` whenever you `in_kernel_invocation_manager`? Can we combine the context managers together then?

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -417,18 +431,17 @@ def nyi(fake_mode, func, *args, **kwargs):
+    prev_in_kernel = fake_mode.in_kernel_invocation
     meta_in_tls = torch._C._meta_in_tls_dispatch_include()
-    prev = fake_mode.in_kernel_invocation
+    assert meta_in_tls == prev_in_kernel, f"{meta_in_tls}, {prev_in_kernel}"
```
If these two pieces of state have to always be synchronized, why not just do away with `fake_mode.in_kernel_invocation` and only use the TLS instead?

eellison replied:
This has already proven useful in debugging multi-threaded inference where we got into a weird state. It is possible for us to get into a weird state in the future with multiple fake tensor modes or something along those lines. Would prefer the separate properties and assertion.

ezyang replied:
What was the resolution for multithreaded inference? If the answer was "we can't do a global state and need TLS", only using the TLS would have prevented the problem in the first place, no?

ezyang replied:
Actually, IIUC, you resolved the problem by disabling multithreading, is that right? But actually, that's papering over the root cause of this problem, which is that in kernel invocation cannot be a per Fake Tensor object property!

ezyang replied:
Like, I think disabling multithreading is still a valid thing to do, but you should still fix the root cause of the bug.

eellison replied:
> which is that in kernel invocation cannot be a per Fake Tensor object property
It's a property on the FakeTensorMode, not on the fake tensor. The problem was concurrent execution of `__torch_dispatch__` where one thread was in kernel invocation so the other thread reported "meta" when it shouldn't have.   

> only using the TLS would have prevented the problem in the first place, no?

Hmm maybe, but there are other potential data races, e.g. in the MetaConverter for preserving aliasing. I would rather not go into the rabbit hole of trying to make Fake Tensor Mode thread safe, especially since it's going to be really hard to test and validate. 

I think the right fix is to have a lock on FakeTensorMode only allowing one thread at a time invoke `__torch_dispatch__`.

ezyang replied:
> It's a property on the FakeTensorMode, not on the fake tensor. The problem was concurrent execution of __torch_dispatch__ where one thread was in kernel invocation so the other thread reported "meta" when it shouldn't have.

Sorry, I wasn't clear. You should have a *per mode* TLS. This can be done by allocating a threading.local per fake tensor

> Hmm maybe, but there are other potential data races, e.g. in the MetaConverter for preserving aliasing. I would rather not go into the rabbit hole of trying to make Fake Tensor Mode thread safe, especially since it's going to be really hard to test and validate.

You cannot have a true data race, because fake tensor is written in Python and for Python code to execute the GIL must be held. You can only have "logical" races, where you do some change and then release the GIL (by calling into PyTorch) allowing another thread to observe the problem. 

A lock around the entire shebang will certainly work, but if you have the lock have a large extent you have to be more worried about accidentally deadlocking (in particular, if you `with self` inside the fake tensor mode, you had better release the lock). I'd prefer a smaller extent lock; e.g., lock FakeConverter, and use TLS for in kernel invocation.

eellison replied:
> You can only have "logical" races, where you do some change and then release the GIL (by calling into PyTorch) allowing another thread to observe the problem.

Yea, what I meant was two threads concurrently check `get_tensor_memo`, and return two separate storages, as opposed to data corruption or something.

> in particular, if you with self inside the fake tensor mode, you had better release the lock

The idea was to only allow one thread access at a time, so this wouldn't be an issue. But a FakeConverter lock and TLS for in kernel invocation could also work.

I agree would be good to fix multi-threading, but at least for PT2, not a critical issue, so will do in follow up.

ezyang reviewed (dismissed):
Approving in the spirit of unblocking, but this is not really the end of the discussion

ezyang commented:
K I'll bounce my review for now

eellison committed:
Update on "Enable aten-aten decomps"


Invokes aten-aten decomps with re-entrant FakeMode. These decomps are being used in other places, so it's good to unify the path static fake tensor takes / get additional testing etc. There is also an instance where we return different devices with cpu/cuda which this fixes ([batch_norm](https://github.com/pytorch/pytorch/blob/master/torch/_decomp/decompositions.py#L1374))



[ghstack-poisoned]

eellison committed:
Update on "Enable aten-aten decomps"


Invokes aten-aten decomps with re-entrant FakeMode. These decomps are being used in other places, so it's good to unify the path static fake tensor takes / get additional testing etc. There is also an instance where we return different devices with cpu/cuda which this fixes ([batch_norm](https://github.com/pytorch/pytorch/blob/master/torch/_decomp/decompositions.py#L1374))



[ghstack-poisoned]

ezyang commented on torch/_decomp/__init__.py:
```diff
@@ -18,6 +18,10 @@
 meta_lib = torch.library.Library("aten", "IMPL", "Meta")
 
+# decompositions which have been disabled as meta kernel implementations,
+# usually due to mismatching strides, aliasing, or other inconsistent property
```
@SherlockNoMad @Chillee this is something we're going to have to resolve before we can move to the single global table... we can't have wrong implementations

ezyang commented on torch/_decomp/decompositions.py:
```diff
@@ -1282,12 +1282,8 @@ def native_layer_norm_backward(
-            if output_mask[2] and bias_cast
-            else None,
+            input.new_zeros(input_shape[axis:]) if output_mask[1] else None,
+            input.new_zeros(input_shape[axis:]) if output_mask[2] else None,
```
I'm still not sure what's going on with this decomp change.

eellison replied:
It was fixing the number of tensors returned, which was caught by tests

ezyang commented on torch/_refs/__init__.py:
```diff
@@ -3525,7 +3525,7 @@ def transpose(a: TensorLikeType, dim0: int, dim1: int) -> TensorLikeType:
 
 
-@register_decomposition(torch.ops.aten.unfold)
+@register_decomposition(torch.ops.aten.unfold, disable_meta=True)
```
can we say what's wrong about this decomposition?

eellison replied:
being fixed in https://github.com/pytorch/pytorch/pull/86428

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -329,7 +344,7 @@ def to_copy(fake_mode, func, *args, **kwargs):
     input_device = new_kwargs.pop("device", None)
     out_device = input_device if input_device else new_kwargs["input"].device
-    with no_dispatch(), in_kernel_invocation_manager(fake_mode):
+    with in_kernel_invocation_manager(fake_mode):
```
What's going on here?

eellison replied:
This is because I moved no_dispatch within `in_kernel_invocation_manager`, as you suggested

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -759,6 +776,15 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+        if (
+            func in decomposition_table
+            and aten_to_aten_decomp(func)
+            and func not in _disabled_meta_decomps
```
@Chillee @SherlockNoMad not sure exactly how this plays in the new one global registration table world order

ezyang reviewed (approved):
Approving as this is directionally good, but I do think there are some questions to still answer.

eellison committed:
Update on "Enable aten-aten decomps"


Invokes aten-aten decomps with re-entrant FakeMode. These decomps are being used in other places, so it's good to unify the path static fake tensor takes / get additional testing etc. There is also an instance where we return different devices with cpu/cuda which this fixes ([batch_norm](https://github.com/pytorch/pytorch/blob/master/torch/_decomp/decompositions.py#L1374))



[ghstack-poisoned]

eellison committed:
Update on "Enable aten-aten decomps"


Invokes aten-aten decomps with re-entrant FakeMode. These decomps are being used in other places, so it's good to unify the path static fake tensor takes / get additional testing etc. There is also an instance where we return different devices with cpu/cuda which this fixes ([batch_norm](https://github.com/pytorch/pytorch/blob/master/torch/_decomp/decompositions.py#L1374))



[ghstack-poisoned]

eellison committed:
Update on "Enable aten-aten decomps"


Invokes aten-aten decomps with re-entrant FakeMode. These decomps are being used in other places, so it's good to unify the path static fake tensor takes / get additional testing etc. There is also an instance where we return different devices with cpu/cuda which this fixes ([batch_norm](https://github.com/pytorch/pytorch/blob/master/torch/_decomp/decompositions.py#L1374))



[ghstack-poisoned]

eellison committed:
Update on "Enable aten-aten decomps"


Invokes aten-aten decomps with re-entrant FakeMode. These decomps are being used in other places, so it's good to unify the path static fake tensor takes / get additional testing etc. There is also an instance where we return different devices with cpu/cuda which this fixes ([batch_norm](https://github.com/pytorch/pytorch/blob/master/torch/_decomp/decompositions.py#L1374))



[ghstack-poisoned]

eellison commented:
@pytorchbot merge

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your PR. I think it breaks a dynamo test in trunk https://hud.pytorch.org/pytorch/pytorch/commit/62e4f51efdf98a3a91d29efa55e5665d5398b464' -c landrace

This looks like a land race

malfet commented:
@pytorchbot merge -f "Looks like false positive, revert did not help"
